### PR TITLE
Refactor memory in Droidlet

### LINKED ISSUE #982: refactor memory
## Type of Issue

Select the type of issue:
- [ ] Bug report (to report a bug)
- [ ] Feature request (to request an additional feature)
- [x] Tracker (I am just using this as a tracker)
- [x] Refactor request
- [ ] Documentation Ask

## Description

There are a large number of methods in the main memory that should be moved to be class methods of the memory nodes.  The behavior of the methods should not change, just their location in the codebase.  If the memory node files get too big, we can split these up.  When this is done, none of the memory-node specific methods (e.g. ~line 370 to line 940 in sql_memory.py) should reside in sql_memory; ditto for mc_memory.py etc.   In addition, many of these methods are overly specific, and should be removed altogether.

We should register the memory nodes classes in the main memory, and continue to allow using the methods by agent_memory.nodes[node_key].method(agent_memory, *args, **args)

## Checklist

Move to class methods of memory nodes:
-[ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L412
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L430
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L453
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L472
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L500
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L590
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L600
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L609
- [ ]  https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L656

to delete:
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L370
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L573
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L582
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L590
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L635
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L643
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L669
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L684
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L694
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L706
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L714
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L726
- [ ] https://github.com/facebookresearch/fairo/blob/8fcc711866af63ee5dd1839a0d356f2208ad54d5/droidlet/memory/sql_memory.py#L734

### PR DESCRIPTION
**Pull request by zephirefaith**
# Description

Refactoring Droidlet agent's memory to make memory access node-centric and deleting over-specific search functions.

This PR deletes the methods referred to in issue #982 with the exception of those which query for MemBlocks based on their specific x,y,z. The generic methods that we have only allow querying wrt avg x,y,z. The deleted methods have been replaced with more generic methods with the same functionality (mostly basic_search).

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [x] I want a high level review. 
- [ ] I want a deep design review.

# Testing

Verified through CircleCI also by running memory-level tests locally

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

**Comment by zephirefaith** [comment]
I am working on refactoring the rest of the listed functions as per the initial snippets above.

**Review by kavyasrinet — APPROVED** [review]
The refactor looks good to me. I would like to add a request to add unit tests for methods that this PR will eventually add - if this PR gets too big for that, lets create an issue that tracks a list of such tests.

**Comment by zephirefaith** [comment]
Thanks for the look-over! The PR will move methods from sql_memory or craftassist memory to specific memory_nodes. Does that mean these methods are not already being tested? If not, is `test_low_level_memory` a good base template to use for building the rest of unit-tests? I will prefer to keep a running list of missing UTs in another issue, so that I can still use the original issue relating to this PR for tracking refactor progress.

**Comment by zephirefaith** [comment]
@kavyasrinet @aszlam I finished the refactor for moving methods out of `sql_memory.py` and into `memory_nodes.py`. The local tests seem a little uneven and I had to dig arbitrarily to figure out which similarly named methods were a part of refactor or not; so history is a little messy, we can clean that up if things look right.  

PS: I might add another task to document and stream-line local testing for contributors. I missed having those instructions/documentation somewhere.

File: agents/craftassist/tests/fake_agent.py [hunk 1]
```diff
@@ -313,13 +313,13 @@ def add_object(
...
+        memid = self.memory.get_object_info_by_xyz(abs_xyz, "BlockObjects")[0]
         for pred, obj in relations.items():
-            self.memory.add_triple(subj=memid, pred_text=pred, obj_text=obj)
+            self.memory.nodes["Triple"].create(self.memory, subj=memid, pred_text=pred, obj_text=obj)
             # sooooorrry  FIXME? when we handle triples better in interpreter_helper
             if "has_" in pred:
-                self.memory.tag(memid, obj)
-        return self.memory.get_object_by_id(memid)
+                self.memory.nodes["Triple"].tag(self.memory, memid, obj)
+        return self.memory.basic_search(f"SELECT MEMORY FROM ReferenceObject WHERE uuid={memid}")[1][0]
```
Comment (aszlam) [inline agents/craftassist/tests/fake_agent.py:322]: can just do memory.get_mem_by_id here, no?

**Review by ethan-carlson — CHANGES_REQUESTED** [review]
Thanks for this PR!  Things look good in general from my perspective, but there are some old memory method calls outstanding that were probably introduced in the most recent rebase, eg. lines 301-310 in dialogue_task.py, and in reference_object_clarification.py

File: agents/craftassist/tests/test_memory.py [hunk 2]
```diff
@@ -72,15 +73,15 @@ def test_add_mob(self):
         # add mob
         chicken = {v: k for k, v in MOBS_BY_ID.items()}["chicken"]
         mob_id, mob_type, pos, look = 42, chicken, Pos(3, 4, 5), Look(0.0, 0.0)
-        self.memory.set_mob_position(Mob(mob_id, mob_type, pos, look))
+        self.memory.nodes[MobNode.NODE_TYPE].set_mob_position(self.memory, Mob(mob_id, mob_type, pos, look))
```
Comment (kavyasrinet) [inline agents/craftassist/tests/test_memory.py:76]: We are using `MobNode.NODE_TYPE` here and `"Triple"`, `"Chat"`, `"Program"` throughout the PR in other places, I propose that we be consistent one way or another. Let's stick with `MemoryNode.NODE_TYPE`

File: agents/droidlet_agent.py [hunk 3]
```diff
@@ -155,8 +155,8 @@ def get_chat_action_dict(sid, chat):
                 chat_memids, _ = self.memory.basic_search(
                     f"SELECT MEMORY FROM Chat WHERE chat={chat}"
                 )
-                logical_form_triples = self.memory.get_triples(
-                    subj=chat_memids[0], pred_text="has_logical_form"
+                logical_form_triples = self.memory.nodes["Triple"].get_triples(
```
Comment (kavyasrinet) [inline agents/droidlet_agent.py:158]: same comment as above

File: droidlet/memory/craftassist/mc_memory_nodes.py [hunk 4]
```diff
@@ -165,13 +268,13 @@ def create(cls, memory, blocks: Sequence[Block]) -> str:
...
         memory.db_write(cmd, memid, 0, 0, 0, "BlockObjects", 0)
         for block in blocks:
-            memory.upsert_block(block, memid, "BlockObjects")
-        memory.tag(memid, "_block_object")
-        memory.tag(memid, "_VOXEL_OBJECT")
-        memory.tag(memid, "_physical_object")
-        memory.tag(memid, "_destructible")
+            VoxelObjectNode.upsert_block(memory, block, memid, "BlockObjects")
+        memory.nodes["Triple"].tag(memory, memid, "_block_object")
+        memory.nodes["Triple"].tag(memory, memid, "VOXEL_OBJECT")
```
Comment (kavyasrinet) [inline droidlet/memory/craftassist/mc_memory_nodes.py:273]: a nit that this should be `_VOXEL_OBJECT`, we use `_`s to represent internal tags

File: droidlet/memory/craftassist/mc_memory_nodes.py [hunk 5]
```diff
@@ -253,25 +356,25 @@ def create(cls, memory, locs, tags=[]) -> str:
         for loc in locs:
             cmd = "INSERT INTO VoxelObjects (uuid, x, y, z, ref_type) VALUES ( ?, ?, ?, ?, ?)"
             memory.db_write(cmd, memid, loc[0], loc[1], loc[2], "inst_seg")
-        memory.tag(memid, "_VOXEL_OBJECT")
-        memory.tag(memid, "_inst_seg")
-        memory.tag(memid, "_destructible")
+        memory.nodes["Triple"].tag(memory, memid, "VOXEL_OBJECT")
```
Comment (kavyasrinet) [inline droidlet/memory/craftassist/mc_memory_nodes.py:359]: same as above

**Review by kavyasrinet — APPROVED** [review]
Thanks for hammering through the codebase and bringing in this change. I like the change, flagged an inconsistency that should be fixed, but other than that, LGTM.

**Review by kavyasrinet — COMMENTED** [review]
Thanks for hammering through the codebase and bringing in this change. I like the change, flagged an inconsistency that should be fixed, but other than that, LGTM.

File: droidlet/memory/craftassist/mc_memory_nodes.py (hunk 4 above)
Comment (zephirefaith) [reply droidlet/memory/craftassist/mc_memory_nodes.py:273]: Thanks so much for catching this! This was the reason Voxel related tests were failing in my branch!

**Comment by zephirefaith** [comment]
@ethan-carlson @kavyasrinet Thanks for your review! I was able to fix the inconsistency as well as the failing tests from those pointers. The inconsistency changes are spread out over all files touched, but local for other fixes. If things look good, then we can merge this in.

**Review by ethan-carlson — APPROVED** [review]
LGTM, thanks!

**Pull request merged**